File: Answer/1291.顺次数.java
/*
 * @lc app=leetcode.cn id=1291 lang=java
 *
 * [1291] 顺次数
 */

import java.util.ArrayList;
import java.util.List;

// @lc code=start
class Solution {

    public List<Integer> sequentialDigits(int low, int high) {
        List<Integer> res = new ArrayList<>(36);
        int delta = 1;
        int offset = 0;
        for (int length = 2; length <= 9; length++) {
            delta = delta * 10 + 1;
            offset = offset * 10 + (length - 1);

            int minNum = delta + offset;
            if (minNum > high) {
                break;
            }
            int maxNum = (10 - length) * delta + offset;
            if (maxNum < low) {
                continue;
            }
            for (int start = 1; start <= 10 - length; start++) {
                int num = start * delta + offset;
                if (num < low) {
                    continue;
                }
                if (num > high) {
                    break;
                }
                res.add(num);
            }
        }
        return res;
    }
}
// @lc code=end

File: Answer/1291.顺次数.py
#
# @lc app=leetcode.cn id=1291 lang=python3
#
# [1291] 顺次数
#

# @lc code=start
class Solution:
    def sequentialDigits(self, low: int, high: int) -> List[int]:
        res = []
        delta = 1
        offset = 0
        for length in range(2, 10):
            delta = delta * 10 + 1
            offset = offset * 10 + (length - 1)

            min_num = delta + offset
            if min_num > high:
                break
            max_num = (10 - length) * delta + offset
            if max_num < low:
                continue
            for start in range(1, 11 - length):
                num = start * delta + offset
                if num < low:
                    continue
                if num > high:
                    break
                res.append(num)
        return res
# @lc code=end
